module/x/gravity/keeper/grpc_query.go

```diff
@@ -1,354 +1,407 @@
+	if !ok {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to contract call for %d %s", req.InvalidationNonce, req.InvalidationScope)
+	}
+
+	// TODO: figure out how to call latest
+
+	return &types.ContractCallTxResponse{LogicCall: cctx}, nil
+}
+
+func (k Keeper) SignerSetTxs(c context.Context, req *types.SignerSetTxsRequest) (*types.SignerSetTxsResponse, error) {
+	var signers []*types.SignerSetTx
+	k.IterateOutgoingTxsByType(sdk.UnwrapSDKContext(c), types.SignerSetTxPrefixByte, func(_ []byte, otx types.OutgoingTx) bool {
+		signer, ok := otx.(*types.SignerSetTx)
+		if !ok {
+			panic(sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to signer set for %s", otx))
+		}
+		signers = append(signers, signer)
+
+		return len(signers) < int(req.Count)
+	})
+	return &types.SignerSetTxsResponse{SignerSets: signers}, nil
+}
+
+func (k Keeper) BatchTxs(c context.Context, req *types.BatchTxsRequest) (*types.BatchTxsResponse, error) {
+	var batches []*types.BatchTx
+	k.IterateOutgoingTxsByType(sdk.UnwrapSDKContext(c), types.BatchTxPrefixByte, func(_ []byte, otx types.OutgoingTx) bool {
+		batch, ok := otx.(*types.BatchTx)
+		if !ok {
+			panic(sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to batch tx for %s", otx))
+		}
+		batches = append(batches, batch)
+		return false
+	})
+	return &types.BatchTxsResponse{Batches: batches}, nil
+}
+
+func (k Keeper) ContractCallTxs(c context.Context, req *types.ContractCallTxsRequest) (*types.ContractCallTxsResponse, error) {
+	var calls []*types.ContractCallTx
+	k.IterateOutgoingTxsByType(sdk.UnwrapSDKContext(c), types.ContractCallTxPrefixByte, func(_ []byte, otx types.OutgoingTx) bool {
+		call, ok := otx.(*types.ContractCallTx)
+		if !ok {
+			panic(sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to contract call for %s", otx))
+		}
+		calls = append(calls, call)
+		return false
+	})
+	return &types.ContractCallTxsResponse{Calls: calls}, nil
+}
+
+func (k Keeper) SignerSetTxEthereumSignatures(c context.Context, req *types.SignerSetTxEthereumSignaturesRequest) (*types.SignerSetTxEthereumSignaturesResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	key := types.MakeSignerSetTxKey(req.Nonce)
+	if req.Address == "" {
+		val, err := k.getSignerValidator(ctx, req.Address)
+		if err != nil {
+			return nil, err
+		}
+		return &types.SignerSetTxEthereumSignaturesResponse{Signature: [][]byte{k.GetEthereumSignature(ctx, key, val)}}, nil
+	}
+
+	var out [][]byte
+	k.IterateEthereumSignatures(ctx, key, func(_ sdk.ValAddress, sig hexutil.Bytes) bool {
+		out = append(out, sig)
+		return false
+	})
+	return &types.SignerSetTxEthereumSignaturesResponse{Signature: out}, nil
+}
+
+func (k Keeper) BatchTxEthereumSignatures(c context.Context, req *types.BatchTxEthereumSignaturesRequest) (*types.BatchTxEthereumSignaturesResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	key := types.MakeBatchTxKey(common.HexToAddress(req.ContractAddress), req.Nonce)
+	if req.Address == "" {
```

Comment:
validate address?

Reply:
It will get validated on the following line